integration/travis/run_integration.sh

```diff
@@ -78,18 +78,36 @@ fi
 
 # Install the CLI
 cd ${PROJECT_DIR}/../cli
-python3 --version
-python3 setup.py install
+python --version
+pip install -e .
 CLI=$(pyenv which cs)
 export PATH=${PATH}:$(dirname ${CLI})
 cs --help
 
 # Run the integration tests
+# We use pytest's --lf option to rerun only the tests that failed previously
+# on each iteration of the loop. All tests are always run on the first iteration.
+# This should help us avoid the need to rerun the whole suite due to a flakey test.
+# We can also scrape the Travis logs to find which tests frequently flake.
 cd ${PROJECT_DIR}
-COOK_MULTI_CLUSTER= COOK_MASTER_SLAVE= COOK_SLAVE_URL=http://localhost:12322 python3 setup.py nosetests --attr ${NOSE_ATTRIBUTES} --verbosity=3 || test_failures=true
+export COOK_MULTI_CLUSTER=
+export COOK_MASTER_SLAVE=
+export COOK_SLAVE_URL=http://localhost:12322
+for ((i=1; i<5; i++)); do
+    pytest -n4 -v --lf --color=no --timeout-method=thread --boxed -m "${PYTEST_MARKS}" && break
+    test_failures=true
+    failure_file="$(find .cache -name lastfailed)"
+    failure_count=$(tail -n+2 $failure_file | wc -l)
+    if [[ $failure_count > 3 ]]; then
+        echo "Too many failures ($failure_count), not retrying..."
+        break
+    fi
+    echo "Retrying tests due to failures (attempt $i)..."
+done
```

Comment:
I'm on the fence about this. On the one hand, I like the idea of making flakes less disruptive. On the other hand, I like the idea of leaving flakes disruptive as motivation to fix the flakiness. I would vote to have this in a subsequent PR, since it's orthogonal from the rest of this PR, I think.

Reply:
I'm not convinced this is a good feature, in any case I agree we should leave it out of this PR.

Reply:
That's a good argument, and I'd be good with doing that. I hinted in the comment that I think we should have another script running to scrape the log files and get the stats on the failures. Then we can collate the stats periodically and see which tests are problematic, but without having to constantly retry jobs in Travis... I actually added that in this PR specifically because tests kept flaking and it was getting irritating...

@pschorf - Thoughts?